pkg/schedule/affinity/policy.go

```diff
@@ -0,0 +1,232 @@
+	interval := getAvailabilityCheckInterval()
+	ticker := time.NewTicker(interval)
+	go func() {
+		defer logutil.LogPanic()
+		defer ticker.Stop()
+		for {
+			select {
+			case <-m.ctx.Done():
+				log.Info("affinity manager availability check loop stopped")
+				return
+			case <-ticker.C:
+				m.checkStoresAvailability()
+			}
+		}
+	}()
+	log.Info("affinity manager availability check loop started", zap.Duration("interval", interval))
+}
+
+// checkStoresAvailability checks the availability status of stores and invalidates groups with unavailable stores.
+func (m *Manager) checkStoresAvailability() {
+	if !m.IsAvailable() {
+		return
+	}
+	unavailableStores := m.generateUnavailableStores()
+	isUnavailableStoresChanged, groupAvailabilityChanges := m.getGroupAvailabilityChanges(unavailableStores)
+	if isUnavailableStoresChanged {
+		m.setGroupAvailabilityChanges(unavailableStores, groupAvailabilityChanges)
+	}
+	m.collectMetrics()
+}
+
+// collectMetrics collects the global metrics of the affinity manager.
+func (m *Manager) collectMetrics() {
+	m.RLock()
+	defer m.RUnlock()
+
+	// Collect global metrics
+	groupCount.Set(float64(len(m.groups)))
+	regionCount.Set(float64(len(m.regions)))
+	affinityRegionCount.Set(float64(m.affinityRegionCount))
+}
+
+func (m *Manager) generateUnavailableStores() map[uint64]storeCondition {
+	unavailableStores := make(map[uint64]storeCondition)
+	stores := m.storeSetInformer.GetStores()
+	lowSpaceRatio := m.conf.GetLowSpaceRatio()
+	for _, store := range stores {
+		switch {
+		// First the conditions that will mark the group as expired
+		case store.IsRemoved() || store.IsPhysicallyDestroyed() || store.IsRemoving():
+			unavailableStores[store.GetID()] = storeRemovingOrRemoved
+		case store.IsUnhealthy():
+			unavailableStores[store.GetID()] = storeDown
+
+		// Then the conditions that will mark the group as degraded
+		case !store.AllowLeaderTransferIn() || m.conf.CheckLabelProperty(config.RejectLeader, store.GetLabels()):
+			unavailableStores[store.GetID()] = storeEvictLeader
+		case store.IsDisconnected():
+			unavailableStores[store.GetID()] = storeDisconnected
+		case store.IsLowSpace(lowSpaceRatio):
+			unavailableStores[store.GetID()] = storeLowSpace
+		case store.IsPreparing():
+			unavailableStores[store.GetID()] = storePreparing
+		}
+		// Note: We intentionally do NOT check:
+		// - IsSlow(): Performance issue, not availability issue
+	}
+	return unavailableStores
+}
+
+func (m *Manager) getGroupAvailabilityChanges(unavailableStores map[uint64]storeCondition) (isUnavailableStoresChanged bool, groupAvailabilityChanges map[string]groupAvailability) {
+	var logEntries []logEntry
+	groupAvailabilityChanges = make(map[string]groupAvailability)
+
+	// Validate whether unavailableStores has changed.
+	m.RLock()
+	isUnavailableStoresChanged = !maps.Equal(unavailableStores, m.unavailableStores)
+	if !isUnavailableStoresChanged {
+		m.RUnlock()
+		return false, nil
```

Comment:
When the store remains unchanged, the group will persist in a degraded state without transitioning to expired.

Reply:
We always use `GetAvailability` to retrieve the availability, which contains the relevant processing logic. We don't update it using `GetAvailability` because it might be called within a read lock.